# Project Team Members

| Prepared by | Email | Prepared for |
| :-: | :-: | :-: |
| **_Hardani Ismu Nabil_** | _hardani.ismu.n@gmail.com_ | **_ND Stadium Collaborative Project 2025/2026_** |

# Notebook Target Definition

* **GENERATES** radio parameters for deployment characterization/analysis

# ===============================
# ===============================

# A- Notebook Setup

In [1]:
# %pip install pandas numpy matplotlib pathlib

In [2]:
import os
import pathlib

import pandas as pd

In [3]:
PROJECT_ROOT = pathlib.Path('.').cwd().parent # you can also do it manually, for example ~/repos/NdSta2026


print('='*50, '[DATA DIRECTORY]')
DATA_DIR2 = PROJECT_ROOT / "data_raw/ND_STA_2026"

print(f"Project Root: {PROJECT_ROOT}")
print(f"Looking for data in: {DATA_DIR2}")

print('='*50, '[DATA DIRECTORY]')
DATA_DIR3 = PROJECT_ROOT / "data_raw/ND_STA_2025"

print(f"Project Root: {PROJECT_ROOT}")
print(f"Looking for data in: {DATA_DIR3}")


print('- '*25)
print('='*50, '[OUTPUT DIRECTORY]')

# OUTPUT_DIR = PROJECT_ROOT / 'output/output_June16th/radio_PciAndRanDist' # CHANGE this into the notebook file name
OUTPUT_DIR = PROJECT_ROOT / 'figures'
os.makedirs(OUTPUT_DIR, exist_ok=True)

print(f"Project Root: {PROJECT_ROOT}")
print(f"All outputs will be in: {OUTPUT_DIR}")

================================================== [DATA DIRECTORY]
Project Root: /home/analisis/repos/NDKI_01_ForGithub_ForArtifact_wintech_2026_stadium_paper_analysis
Looking for data in: /home/analisis/repos/NDKI_01_ForGithub_ForArtifact_wintech_2026_stadium_paper_analysis/data_raw/ND_STA_2026
================================================== [DATA DIRECTORY]
Project Root: /home/analisis/repos/NDKI_01_ForGithub_ForArtifact_wintech_2026_stadium_paper_analysis
Looking for data in: /home/analisis/repos/NDKI_01_ForGithub_ForArtifact_wintech_2026_stadium_paper_analysis/data_raw/ND_STA_2025
- - - - - - - - - - - - - - - - - - - - - - - - - 
================================================== [OUTPUT DIRECTORY]
Project Root: /home/analisis/repos/NDKI_01_ForGithub_ForArtifact_wintech_2026_stadium_paper_analysis
All outputs will be in: /home/analisis/repos/NDKI_01_ForGithub_ForArtifact_wintech_2026_stadium_paper_analysis/figures


In [4]:
lte_radio26 = pd.read_csv(DATA_DIR2 / "lte_radio.csv")
nr_radio26 = pd.read_csv(DATA_DIR2 / "nr_radio.csv")

lte_radio26.rename(columns={'PCI LTE': 'PCI'}, inplace=True)
nr_radio26.rename(columns={'PCI 5G NR': 'PCI'}, inplace=True)
nr_radio26.rename(columns={'SS-RSRP': 'RSRP'}, inplace=True)
nr_radio26.rename(columns={'SS-RSRQ': 'RSRQ'}, inplace=True)
nr_radio26.rename(columns={'SS-SINR': 'SINR'}, inplace=True)

lte_radio25 = pd.read_csv(DATA_DIR3 / "lte_radio.csv")
lte_radio25['Technology'] = 'LTE'
nr_radio25 = pd.read_csv(DATA_DIR3 / "nr_radio.csv")
nr_radio25['Technology'] = '5G NR'

lte_radio25.rename(columns={'PCI LTE': 'PCI'}, inplace=True)
nr_radio25.rename(columns={'PCI 5G NR': 'PCI'}, inplace=True)
nr_radio25.rename(columns={'SS-RSRP': 'RSRP'}, inplace=True)
nr_radio25.rename(columns={'SS-RSRQ': 'RSRQ'}, inplace=True)
nr_radio25.rename(columns={'SS-SINR': 'SINR'}, inplace=True)

/tmp/ipykernel_29403/4015034248.py:10: DtypeWarning: Columns (0: Direction) have mixed types. Specify dtype option on import or set low_memory=False.
  lte_radio25 = pd.read_csv(DATA_DIR3 / "lte_radio.csv")


## B- Data Filtering
by Collection column

In [5]:
COLLECTION_2025_GAME = [
    '20250913_Game_Bowl', '20250920_Game_Bowl', 
    '20251108_Game_Bowl', '20251122_Game_Bowl',
    # '20251004_Game_Bowl'
] 
COLLECTION_2026_GAME = []

COLLECTION_2025_EMPTY = ['20250826_preGame_Bowl', '20250828_preGame_Bowl']
COLLECTION_2026_EMPTY = ['20260226_preGame_Bowl']

In [6]:
# Label the browser data
df_lteRadio26_empty = lte_radio26[lte_radio26['Collection'].isin(COLLECTION_2026_EMPTY)].copy()
df_lteRadio26_empty['Label'] = 'Non-game'
df_lteRadio26_empty['Label 2'] = 'Non-game 2026'

# Label the browser data
df_nrRadio26_empty = nr_radio26[nr_radio26['Collection'].isin(COLLECTION_2026_EMPTY)].copy()
df_nrRadio26_empty['Label'] = 'Non-game'
df_nrRadio26_empty['Label 2'] = 'Non-game 2026'

# Label the browser data
df_lteRadio25_empty = lte_radio25[lte_radio25['Collection'].isin(COLLECTION_2025_EMPTY)].copy()
df_lteRadio25_empty['Label'] = 'Non-game'
df_lteRadio25_empty['Label 2'] = 'Non-game 2025'

# Label the browser data
df_nrRadio25_empty = nr_radio25[nr_radio25['Collection'].isin(COLLECTION_2025_EMPTY)].copy()
df_nrRadio25_empty['Label'] = 'Non-game'
df_nrRadio25_empty['Label 2'] = 'Non-game 2025'

lteRadio_empty = pd.concat([df_lteRadio25_empty, df_lteRadio26_empty], ignore_index=True)
nrRadio_empty = pd.concat([df_nrRadio25_empty, df_nrRadio26_empty], ignore_index=True)

# Label the browser data
lteRadio_game = lte_radio25[lte_radio25['Collection'].isin(COLLECTION_2025_GAME)].copy()
lteRadio_game['Label'] = 'Game'
lteRadio_game['Label 2'] = 'Game 2025'

# Label the browser data
nrRadio_game = nr_radio25[nr_radio25['Collection'].isin(COLLECTION_2025_GAME)].copy()
nrRadio_game['Label'] = 'Game'
nrRadio_game['Label 2'] = 'Game 2025'


df_radio = pd.concat([lteRadio_game, nrRadio_game, lteRadio_empty, nrRadio_empty], ignore_index=True)

In [7]:
look_radio = df_radio[
    ~(df_radio['Collection'] == '')
    # & (df_radio['Collection'] == '')
].copy()

# ===============================
# ===============================

# B- RAN Configuration Distribution

**This one is USED in the paper**

In [8]:
import pandas as pd

# 1. Prepare data
df_look_ran = look_radio.copy()
df_look_ran = df_look_ran[df_look_ran['RAN Configuration'] != 'WiFi'].copy()

# 2. Deduplicate safely by taking the mode of 'RAN Configuration'
df_deduped = (
    df_look_ran.groupby(['Time', 'Label', 'Operator'])['RAN Configuration']
    .agg(lambda x: m[0] if len(m := x.mode()) > 0 else None)
    .reset_index()
)

# 3. Get value counts from the deduplicated data
counts = (
    df_deduped[['Label', 'Operator', 'RAN Configuration']]
    .value_counts(dropna=True)
    .sort_index()
)

# 4. Calculate percentages within each ('Label', 'Operator') group
percentages = counts.groupby(level=['Label', 'Operator']).transform(
    lambda x: round(x / x.sum() * 100, 2)
)

# 5. Combine into final dataframe and display
result = pd.DataFrame({'Count': counts, 'Percentage (%)': percentages})
print(result)

                                     Count  Percentage (%)
Label    Operator RAN Configuration                       
Game     AT&T     5G EN-DC           21313           56.78
                  LTE                16220           43.22
         T-Mobile 5G EN-DC           12816           42.96
                  5G SA              14132           47.38
                  LTE                 2882            9.66
         Verizon  5G EN-DC           15799           63.90
                  LTE                 8925           36.10
Non-game AT&T     5G EN-DC           15267           58.59
                  5G SA                  1            0.00
                  LTE                10791           41.41
         T-Mobile 5G SA              13134           96.21
                  LTE                  517            3.79
         Verizon  5G EN-DC            7461           39.91
                  LTE                11235           60.09


### B: Whatsapp/Instagram specific

In [9]:
import pandas as pd

# 1. Prepare data
df_look_ran = look_radio[look_radio['Test Name'] == 'Instagram'].copy()
# df_look_ran = look_radio.copy()

# 2. Deduplicate safely by taking the mode of 'RAN Configuration'
df_deduped = (
    df_look_ran.groupby(['Time', 'Label', 'Operator'])[['RAN Configuration', 'Band Number']]
    .agg(lambda x: m[0] if len(m := x.mode()) > 0 else None)
    .reset_index()
)

# 3. Get value counts from the deduplicated data
counts = (
    df_deduped[['Label', 'Operator', 'RAN Configuration', 'Band Number']]
    .value_counts(dropna=True)
    .sort_index()
)

# 4. Calculate percentages within each ('Label', 'Operator') group
percentages = counts.groupby(level=['Label', 'Operator']).transform(
    lambda x: round(x / x.sum() * 100, 2)
)

# 5. Combine into final dataframe and display
result = pd.DataFrame({'Count': counts, 'Percentage (%)': percentages})
print(result)

                                                 Count  Percentage (%)
Label    Operator RAN Configuration Band Number                       
Game     AT&T     5G EN-DC          2.0            206           20.54
                                    5.0            264           26.32
                                    30.0            61            6.08
                                    66.0             5            0.50
                  LTE               2.0            218           21.73
                                    12.0            34            3.39
                                    30.0            93            9.27
                                    66.0           122           12.16
         T-Mobile 5G EN-DC          25.0            10            1.67
                  5G SA             25.0           116           19.33
                                    41.0           267           44.50
                                    66.0             1            0.17
      

In [10]:
import pandas as pd

# 1. Prepare data
df_look_ran = look_radio[look_radio['Test Name'] == 'WhatsApp Send'].copy()
# df_look_ran = look_radio.copy()

# 2. Deduplicate safely by taking the mode of 'RAN Configuration'
df_deduped = (
    df_look_ran.groupby(['Time', 'Label', 'Operator'])[['RAN Configuration', 'Band Number']]
    .agg(lambda x: m[0] if len(m := x.mode()) > 0 else None)
    .reset_index()
)

# 3. Get value counts from the deduplicated data
counts = (
    df_deduped[['Label', 'Operator', 'RAN Configuration', 'Band Number']]
    .value_counts(dropna=True)
    .sort_index()
)

# 4. Calculate percentages within each ('Label', 'Operator') group
percentages = counts.groupby(level=['Label', 'Operator']).transform(
    lambda x: round(x / x.sum() * 100, 2)
)

# 5. Combine into final dataframe and display
result = pd.DataFrame({'Count': counts, 'Percentage (%)': percentages})
print(result)

                                                 Count  Percentage (%)
Label    Operator RAN Configuration Band Number                       
Game     AT&T     5G EN-DC          2.0            412           29.58
                                    5.0            453           32.52
                                    30.0            20            1.44
                                    66.0            18            1.29
                  LTE               2.0            265           19.02
                                    12.0            33            2.37
                                    30.0            92            6.60
                                    66.0           100            7.18
         T-Mobile 5G EN-DC          25.0             4            0.86
                  5G SA             25.0            50           10.78
                                    41.0           302           65.09
                                    66.0             1            0.22
      

# C- Unique PCI

In [11]:
df_look_pci = look_radio[
    ~(look_radio['Collection'] == '')
    # & (look_radio['Test Status'] == 'Completed')
    # & (look_radio['Collection'].isin(COLLECTION_2025_EMPTY))
    # & (look_radio['Collection'].isin(['20250826_preGame_Bowl']))
    # & (look_radio['Operator'].isin(['T-Mobile']))
    # & (look_radio['Band Number'].isin([4, 71]))
    # & (look_radio['Technology'].isin(['LTE']))
].copy()
# df_look_pci = look_radio.copy()

threshold = 0  # Example threshold rows
cols = [
    'Technology',
    'Label', 'Operator', 
    # 'RAN Configuration',
    # 'Band Number'
]

pci_occurrences = df_look_pci.value_counts(subset=cols + ['PCI'])
frequent_pcis = pci_occurrences.loc[lambda x: x >= threshold].reset_index()
final_unique_counts = frequent_pcis.groupby(cols)['PCI'].nunique()
print(final_unique_counts.sort_index().to_string())

# print(df_look_pci[['Collection', 'Operator', 'RAN Configuration', 'Band Number']].value_counts().sort_index().to_string())

Technology  Label     Operator
5G NR       Game      AT&T         26
                      T-Mobile     63
                      Verizon      13
            Non-game  AT&T         21
                      T-Mobile     19
                      Verizon       4
LTE         Game      AT&T        130
                      T-Mobile     48
                      Verizon      36
            Non-game  AT&T         93
                      T-Mobile      4
                      Verizon      23


**USED IN THE PAPER**

    Technology  Label     Operator
    5G NR       Game      AT&T         26
                        T-Mobile     63
                        Verizon      13
                Non-game  AT&T         21
                        T-Mobile     19
                        Verizon       4
    LTE         Game      AT&T        130
                        T-Mobile     48
                        Verizon      36
                Non-game  AT&T         93
                        T-Mobile      4
                        Verizon      23

# D- Bandwidth

In [12]:
df_look_pci = look_radio[
    ~(look_radio['Collection'] == '')
    # & (look_radio['Test Status'] == 'Completed')
    # & (look_radio['Collection'].isin(COLLECTION_2025_EMPTY))
    # & (look_radio['Collection'].isin(['20250826_preGame_Bowl']))
].copy()
# df_look_pci = look_radio.copy()


cols = [
    'Technology',
    'Label', 'Operator', 
    # 'RAN Configuration',
    'Band Number'
]
print(df_look_pci[cols + ['UL Bandwidth', 'DL Bandwidth']].value_counts(dropna=False).sort_index().to_string())

Technology  Label     Operator  Band Number  UL Bandwidth  DL Bandwidth
5G NR       Game      AT&T      5.0          10.0          10.0            10724
                      T-Mobile  25.0         20.0          20.0             3448
                                             NaN           20.0               42
                                41.0         90.0          90.0             6405
                                             100.0         100.0           10637
                                             NaN           45.0                2
                                                           50.0                3
                                                           90.0               28
                                                           100.0              91
                                71.0         20.0          20.0              143
                      Verizon   77.0         100.0         100.0            6659
                                     